pkg/issuer/acme/dns/webhook/webhook.go

```diff
@@ -0,0 +1,131 @@
+)
+
+const (
+	maxBodyLength = 5 * 1024 * 1024
+
+	presentOperation operation = "present"
+	cleanupOperation operation = "cleanup"
+)
+
+type operation string
+
+type WebhookPayload struct {
+	Operation operation `json:"operation"`
+
+	// Identifier is the record that should have a TXT record set for, e.g. _acme-challenge.example.com
+	Identifier string `json:"identifier"`
+
+	// Key is the value that the TXT record should hold for this domain
+	Key string `json:"key"`
+
+	// Metadata is arbitrary additional metadata passed to the plugin from the Issuer resource.
+	// This may contain a reference to a secret resource, containing secret data specific to this
+	// configuration of the plugin.
+	Metadata map[string]string `json:"metadata,omitempty"`
+}
+
+type DNSProvider struct {
+	httpClient       *http.Client
+	url              string
+	metadata         map[string]string
+	dns01Nameservers []string
+}
+
+func NewDNSProvider(url string, metadata map[string]string, skipTLSVerify bool, webhookCA []byte, dns01Nameservers []string) (*DNSProvider, error) {
+	rootCAs, err := x509.SystemCertPool()
+	if rootCAs == nil || err != nil {
+		glog.Errorf("can't instantiate system CA pool, falling back to empty CA pool: %s", err)
+	}
+
+	if ok := rootCAs.AppendCertsFromPEM(webhookCA); !ok {
+		glog.Infof("no certs appended, using system certs only")
+	}
+
+	tlsConfig := &tls.Config{ClientCAs: rootCAs, InsecureSkipVerify: skipTLSVerify}
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+		Timeout: time.Duration(30 * time.Second),
+	}
+
+	return &DNSProvider{
+		httpClient:       client,
+		url:              url,
+		metadata:         metadata,
+		dns01Nameservers: dns01Nameservers,
+	}, nil
+}
+
+func (c *DNSProvider) Present(domain, fqdn, value string) error {
+	body, err := c.sendRequest(WebhookPayload{Operation: presentOperation, Identifier: util.UnFqdn(fqdn), Key: value, Metadata: c.metadata})
+	if len(body) > 0 {
+		glog.Infof("response body: %q", body)
+	}
+
+	return err
+}
+
+func (c *DNSProvider) CleanUp(domain, fqdn, value string) error {
+	body, err := c.sendRequest(WebhookPayload{Operation: cleanupOperation, Identifier: util.UnFqdn(fqdn), Key: value, Metadata: c.metadata})
+	if len(body) > 0 {
+		glog.Infof("response body: %q", body)
+	}
+
+	return err
+}
+
+func (c *DNSProvider) sendRequest(payload WebhookPayload) (responseBody string, err error) {
+	jsonString, err := json.Marshal(payload)
+	if err != nil {
+		return "", err
+	}
+
+	response, err := c.httpClient.Post(c.url, "application/json", bytes.NewReader(jsonString))
+	if err != nil {
+		return "", err
+	}
+	if response.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("non-200 HTTP code returned: %d", response.StatusCode)
+	}
+
+	defer response.Body.Close()
+
+	response.Body = http.MaxBytesReader(nil, response.Body, maxBodyLength)
+	buf, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return fmt.Sprintf("truncated: %s", buf), nil
```

Comment:
Could you update this error message to be clearer? Bare in mind that users will not see much more than `truncated: xyz` in their error logs right now.